File: backend/analyzer/src/main/java/com/mategka/dava/analyzer/serialization/model/IndexRootDto.java
package com.mategka.dava.analyzer.serialization.model;

import com.mategka.dava.analyzer.struct.property.value.Kind;
import com.mategka.dava.analyzer.struct.property.value.Visibility;

import com.google.common.collect.Multimap;
import lombok.Builder;
import lombok.NonNull;
import lombok.Value;
import org.jetbrains.annotations.NotNull;

import java.time.YearMonth;

@Value
@Builder
public class IndexRootDto {

  // Name index is compiled at deserialization since it requires fuzzy search support

  @NonNull
  Multimap<@NotNull Visibility, @NotNull Long> byVisibility;

  @NonNull
  Multimap<@NotNull Kind, @NotNull Long> byKind;

  @NonNull
  Multimap<@NotNull String, @NotNull Long> byType;

  @NonNull
  Multimap<@NotNull YearMonth, @NotNull Long> byExistence;

  @NonNull
  Multimap<@NotNull YearMonth, @NotNull Long> byChanged;

}

File: frontend/src/models/analyzer/IndexRootDto.ts
import type { Visibility } from "@/models/analyzer/value/Visibility"
import type { Kind } from "@/models/analyzer/value/Kind"
import type { QualifiedName, RawYearMonth, SymbolId } from "@/models/analyzer/common"

export type Index<K extends string> = Record<K, SymbolId[]>;

export interface IndexRootDto {
  byVisibility: Index<Visibility>;
  byKind: Index<Kind>;
  byType: Index<QualifiedName>;
  byExistence: Index<RawYearMonth>;
  byChanged: Index<RawYearMonth>;
}
